crates/wysiwyg/src/composer_model/lists.rs

```diff
@@ -357,4 +359,464 @@ where
             }
         }
     }
+
+    pub fn can_indent(&self, locations: Vec<DomLocation>) -> bool {
+        for loc in locations {
+            if loc.is_leaf && !self.can_indent_handle(&loc.node_handle) {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn can_indent_handle(&self, handle: &DomHandle) -> bool {
+        if let DomNode::Container(parent) =
+            self.state.dom.lookup_node(&handle.parent_handle())
+        {
+            let parent = parent;
+
+            // Parent must be a ListItem
+            if parent.is_list_item() {
+                return handle.parent_handle().index_in_parent() > 0;
+            }
+        } else {
+            panic!("Parent node must be a ContainerNode");
+        }
+        false
+    }
+
+    pub fn can_un_indent(&self, locations: Vec<DomLocation>) -> bool {
+        for loc in locations {
+            if loc.is_leaf && !self.can_un_indent_handle(&loc.node_handle) {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn can_un_indent_handle(&self, handle: &DomHandle) -> bool {
+        // Check that there are at least 2 ancestor lists
+        if let Some(closest_list_handle) =
+            self.find_closest_list_ancestor(&handle)
+        {
+            self.find_closest_list_ancestor(&closest_list_handle)
+                .is_some()
+        } else {
+            false
+        }
+    }
+
+    fn find_closest_list_ancestor(
```

Comment:
This and `find_parent_list_item` in `dom_struct.rs` should be near each other. I'm not actually sure where...maybe even in a separate module if we think there will be enough of them?